src/RateLimiter/RateLimiter.php

```diff
@@ -0,0 +1,83 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Yiisoft\Yii\Web\RateLimiter;
+
+use Psr\Http\Message\ResponseFactoryInterface;
+use Psr\Http\Message\ResponseInterface;
+use Psr\Http\Message\ServerRequestInterface;
+use Psr\Http\Server\MiddlewareInterface;
+use Psr\Http\Server\RequestHandlerInterface;
+
+/**
+ * RateLimiter limits the number of consequential requests ({@see CacheCounter::$limit}) that could be processed per
+ * {@see CacheCounter::$interval}. If the number is reached, middleware responds with HTTP code 429, "Too Many Requests"
+ * until limit expires.
+ */
+final class RateLimiter implements MiddlewareInterface
+{
+    private CacheCounter $counter;
```

Comment:
designed for cache interface only?

Reply:
Yes. Do you think you need other storage?

Reply:
The cache (or storage) interface does not support atomically operation.

Reply:
Yes, indeed, its own interface would fit it much better since without atomic operation of incrementing a counter it won't work well under high loads. There will be request spikes.